File: Daily/1266.minimum-time-visiting-all-points.js
/*
 * @lc app=leetcode id=1266 lang=javascript
 *
 * [1266] Minimum Time Visiting All Points
 */

// @lc code=start
/** @typedef {[x: number, y: number]} {Point} */

/**
 * @param {Point[]} points
 * @return {number}
 */
var minTimeToVisitAllPoints = function (points) {
    let minTime = 0;
    for (let i = 0; i < points.length - 1; i++) {
        minTime += minTimeBetween(points[i], points[i + 1]);
    }
    return minTime;
};

/**
 *
 * @param {Point} point1
 * @param {Point} point2
 * @returns {number}
 */
function minTimeBetween(point1, point2) {
    const xDiff = Math.abs(point1[0] - point2[0]);
    const yDiff = Math.abs(point1[1] - point2[1]);

    const diagonalTime = Math.min(xDiff, yDiff);
    return diagonalTime + (xDiff - diagonalTime) + (yDiff - diagonalTime);
}
// @lc code=end

File: src/2026/1266.minimum-time-visiting-all-points.cpp
/*
 * @lc app=leetcode id=1266 lang=cpp
 *
 * [1266] Minimum Time Visiting All Points
 */

#include <vector>

// @lc code=start
class Solution {
 public:
  int minTimeToVisitAllPoints(const std::vector<std::vector<int>>& points) {
    int total_time = 0;
    for (int i = 0; i < points.size() - 1; i++) {
      const std::vector<int>& point1 = points[i];
      const std::vector<int>& point2 = points[i + 1];

      const int x_diff = std::abs(point1[0] - point2[0]);
      const int y_diff = std::abs(point1[1] - point2[1]);

      total_time += std::min(x_diff, y_diff) + std::abs(x_diff - y_diff);
    }
    return total_time;
  }
};
// @lc code=end
